src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs

```diff
@@ -76,20 +76,20 @@ public override BoundNode VisitUnaryOperator(BoundUnaryOperator node)
         {
             if (kind.IsDynamic())
             {
-                Debug.Assert(kind == UnaryOperatorKind.DynamicTrue && type.SpecialType == SpecialType.System_Boolean || type.IsDynamic());
+                Debug.Assert((kind == UnaryOperatorKind.DynamicTrue || kind == UnaryOperatorKind.DynamicFalse) && type.SpecialType == SpecialType.System_Boolean
+                    || type.IsDynamic());
                 Debug.Assert((object)method == null);
 
                 // Logical operators on boxed Boolean constants:
                 var constant = UnboxConstant(loweredOperand);
                 if (constant == ConstantValue.True || constant == ConstantValue.False)
                 {
-                    if (kind == UnaryOperatorKind.DynamicTrue)
+                    switch (kind)
                     {
-                        return _factory.Literal(constant.BooleanValue);
-                    }
-                    else if (kind == UnaryOperatorKind.DynamicLogicalNegation)
-                    {
-                        return MakeConversionNode(_factory.Literal(!constant.BooleanValue), type, @checked: false);
+                        case UnaryOperatorKind.DynamicTrue:
```

Comment:
>case UnaryOperatorKind.DynamicTrue [](start = 24, length = 34)

Why are we not handling UnaryOperatorKind.DynamicFalse by inverting the constant.BooleanValue?

Reply:
There are a few ways to current produce an expression with type dynamic and a constant value true or false, but I didn't manage to a constant from an equality/inequality of two dynamics.
I'll leave as-is.

---
In reply to: [172092212](https://github.com/dotnet/roslyn/pull/23356#discussion_r172092212) [](ancestors = 172092212)